File: crates/driftguard-core/src/evals.rs
//! Eval-case services.
//!
//! Like `prompts`, these are plain async functions over `&PgPool`. Phase 4's
//! fixtures loader will create eval cases in bulk through `create_eval_case`;
//! the CLI also exposes a one-off `eval add` for testing.

use sqlx::PgPool;
use uuid::Uuid;

use crate::error::CoreError;
use crate::models::{EvalCase, EvalRun};
use crate::prompts::get_prompt_by_name;

pub async fn create_eval_case(
    pool: &PgPool,
    prompt_name: &str,
    input: &str,
    expected_behavior: &str,
) -> Result<EvalCase, CoreError> {
    let prompt = get_prompt_by_name(pool, prompt_name)
        .await?
        .ok_or_else(|| CoreError::PromptNotFound(prompt_name.to_string()))?;

    let case = sqlx::query_as!(
        EvalCase,
        r#"INSERT INTO eval_cases (prompt_id, input, expected_behavior)
           VALUES ($1, $2, $3)
           RETURNING id, prompt_id, input, expected_behavior, created_at"#,
        prompt.id,
        input,
        expected_behavior
    )
    .fetch_one(pool)
    .await?;
    Ok(case)
}

pub async fn list_eval_cases(
    pool: &PgPool,
    prompt_id: Uuid,
) -> Result<Vec<EvalCase>, CoreError> {
    let rows = sqlx::query_as!(
        EvalCase,
        r#"SELECT id, prompt_id, input, expected_behavior, created_at
           FROM eval_cases WHERE prompt_id = $1 ORDER BY created_at ASC"#,
        prompt_id
    )
    .fetch_all(pool)
    .await?;
    Ok(rows)
}

/// All eval runs recorded for one prompt version (the read side for the API).
pub async fn list_eval_runs(
    pool: &PgPool,
    version_id: Uuid,
) -> Result<Vec<EvalRun>, CoreError> {
    let rows = sqlx::query_as!(
        EvalRun,
        r#"SELECT id, prompt_version_id, eval_case_id, actual_output,
                  judge_passed, judge_justification, created_at
           FROM eval_runs WHERE prompt_version_id = $1 ORDER BY created_at ASC"#,
        version_id
    )
    .fetch_all(pool)
    .await?;
    Ok(rows)
}

File: internal/core/evals.go
package core

import (
	"context"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
)

// Eval-case services. Read paths only for now; bulk creation arrives with the
// fixtures loader port (Phase G5).

// ListEvalCases returns a prompt's eval cases, oldest first.
func ListEvalCases(ctx context.Context, pool *pgxpool.Pool, promptID uuid.UUID) ([]EvalCase, error) {
	rows, err := pool.Query(ctx,
		`SELECT id, prompt_id, input, expected_behavior, created_at
		 FROM eval_cases WHERE prompt_id = $1 ORDER BY created_at ASC`, promptID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	cases := []EvalCase{}
	for rows.Next() {
		var c EvalCase
		if err := rows.Scan(&c.ID, &c.PromptID, &c.Input, &c.ExpectedBehavior, &c.CreatedAt); err != nil {
			return nil, err
		}
		c.CreatedAt = c.CreatedAt.UTC() // wire contract is UTC (see prompts.go)
		cases = append(cases, c)
	}
	return cases, rows.Err()
}

// ListEvalRuns returns all eval runs recorded for one prompt version (the read
// side for the API).
func ListEvalRuns(ctx context.Context, pool *pgxpool.Pool, versionID uuid.UUID) ([]EvalRun, error) {
	rows, err := pool.Query(ctx,
		`SELECT id, prompt_version_id, eval_case_id, actual_output,
		        judge_passed, judge_justification, created_at
		 FROM eval_runs WHERE prompt_version_id = $1 ORDER BY created_at ASC`, versionID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	runs := []EvalRun{}
	for rows.Next() {
		var r EvalRun
		if err := rows.Scan(&r.ID, &r.PromptVersionID, &r.EvalCaseID, &r.ActualOutput,
			&r.JudgePassed, &r.JudgeJustification, &r.CreatedAt); err != nil {
			return nil, err
		}
		r.CreatedAt = r.CreatedAt.UTC() // wire contract is UTC (see prompts.go)
		runs = append(runs, r)
	}
	return runs, rows.Err()
}

// CreateEvalCase adds one eval case to a prompt by name (used by the fixtures
// loader in bulk; the CLI also exposes a one-off `eval add`).
func CreateEvalCase(ctx context.Context, pool *pgxpool.Pool, promptName, input, expectedBehavior string) (*EvalCase, error) {
	prompt, err := GetPromptByName(ctx, pool, promptName)
	if err != nil {
		return nil, err
	}
	if prompt == nil {
		return nil, PromptNotFoundError{Name: promptName}
	}
	return CreateEvalCaseForPrompt(ctx, pool, prompt.ID, input, expectedBehavior)
}

// CreateEvalCaseForPrompt adds one eval case to a prompt by id. The by-id form
// is what the dashboard write path uses (it already has the prompt id from the
// registry); CreateEvalCase is the by-name convenience over it.
func CreateEvalCaseForPrompt(ctx context.Context, pool *pgxpool.Pool, promptID uuid.UUID, input, expectedBehavior string) (*EvalCase, error) {
	var c EvalCase
	err := pool.QueryRow(ctx,
		`INSERT INTO eval_cases (prompt_id, input, expected_behavior)
		 VALUES ($1, $2, $3)
		 RETURNING id, prompt_id, input, expected_behavior, created_at`,
		promptID, input, expectedBehavior).
		Scan(&c.ID, &c.PromptID, &c.Input, &c.ExpectedBehavior, &c.CreatedAt)
	if err != nil {
		return nil, err
	}
	c.CreatedAt = c.CreatedAt.UTC()
	return &c, nil
}
